src/renderer/apiFetch.ts

```diff
@@ -0,0 +1,80 @@
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import { apiActions } from "readium-desktop/common/redux/actions";
+import { TApiMethod, TApiMethodName } from "readium-desktop/main/api/api.type";
+import { TMethodApi, TModuleApi } from "readium-desktop/main/di";
+import { diRendererGet } from "readium-desktop/renderer/di";
+import { ApiLastSuccess } from "readium-desktop/renderer/redux/states/api";
+import { ReturnPromiseType } from "readium-desktop/typings/promise";
+import { Unsubscribe } from "redux";
+import * as uuid from "uuid";
+
+export async function apiFetch<T extends TApiMethodName>(apiPath: T, ...requestData: Parameters<TApiMethod[T]>) {
+    return new Promise<ReturnPromiseType<TApiMethod[T]>>((resolve, reject) => {
+        const store = diRendererGet("store");
+        const requestId = uuid.v4();
+        const splitPath = apiPath.split("/");
+        const moduleId = splitPath[0] as TModuleApi;
+        const methodId = splitPath[1] as TMethodApi;
+        let lastSuccess: ApiLastSuccess | undefined;
+        let storeUnsubscribe: Unsubscribe| undefined;
+        let timeoutId: number | undefined;
+
+        store.dispatch(
+            apiActions.buildRequestAction(
+                requestId,
+                moduleId,
+                methodId,
+                requestData,
+            ),
+        );
+
+        const promise = new Promise<ReturnPromiseType<TApiMethod[T]>>((resolveSubscribe, rejectSubscribe) => {
+            storeUnsubscribe = store.subscribe(() => {
+                const state = store.getState();
+                const apiLastSuccess = state.api.lastSuccess;
+                const lastSuccessDate = (lastSuccess && lastSuccess.date) || 0;
+
+                if (!apiLastSuccess || apiLastSuccess.date <= lastSuccessDate) {
+                    return;
+                }
+
+                // New api success
+                lastSuccess = apiLastSuccess;
+
+                const meta = apiLastSuccess.action.meta.api;
+                if (moduleId === meta.moduleId && methodId === meta.methodId && state.api.data[requestId]) {
+                    const request = Object.assign({}, state.api.data[requestId]);
+                    store.dispatch(apiActions.clean(requestId));
+                    if (request.resultIsReject) {
+                        rejectSubscribe(request.result);
+                    }
+                    resolveSubscribe(request.result);
+                }
+
+                // handle promise<void>
+                timeoutId = window.setTimeout(() => reject("API Timeout"), 5000);
+            });
+        });
+
+        // The linter doesn't accept .finaly(). Why ? Is it a Bug ?
+        // tslint:disable-next-line: no-floating-promises
```

Comment:
When I delete the comments above, `npm run lint` and `npm run build` complete successfully. Could you please double-check at your end? (prerequisite: make sure to use the latest `package-lock.json` from the develop branch: `rm -rf node_modules && npm install && npm run __postinstall`)

Longer analysis:

Thorium's current Chromium version is '66.0.3359.181', and NodeJS is version `10.2.0` (see `process.versions` debug console output when the app launches).

* In my system shell: `node --version` => `v10.16.3`, and `node -e "Promise.resolve().finally(() => {console.log('OK')})"` => `OK` (see https://node.green/#ES2018-features-Promise-prototype-finally )

* In Thorium's web inspector console (`renderer` process, Chromium V8 runtime): `Promise.resolve().finally(() => {console.log('OK')})` => `OK` (see https://caniuse.com/#feat=promise-finally )

* In Thorium's `main` process, NodeJS V8 runtime: `Promise.resolve().finally(() => { console.log("OK"); });` => `OK` (see https://node.green/#ES2018-features-Promise-prototype-finally )


However, `Promise.finally` is *normally* not included in Thorium's current TypeScript type definitions (ES6/2015 target + ES2017 lib):
https://github.com/readium/readium-desktop/blob/develop/tsconfig.json#L57-L58
Typically, `es2018.promise` needs to be *explicitly* added to the `lib` configuration: `"lib": ["es2017", "es6", "dom", "dom.iterable", "es2018.promise"]`.

So, I do not understand why Thorium compiles successfully given the current `target` and `lib` configuration. I would expect failure because `es2018.promise` is normally not included by default (!?)

Side note - the NodeJS v10 typings also limit a number of features to match the specific NodeJS runtime API of Electron v3:
https://github.com/readium/readium-desktop/blob/54685000999c2802fa5f3a068f2a1f535eff5fd8/package.json#L226

Other references:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

Reply:
i don't know, strange behaviour

b65f342